### PR TITLE
fix(dialog): guard async_ask() against empty or invalid kb_ids

### LINKED ISSUE #15529: [Bug]: async_ask() crashes with IndexError when kb_ids is empty or all IDs are invalid
### Self Checks

- [x] I have searched for existing issues [search for existing issues](https://github.com/infiniflow/ragflow/issues), including closed ones.
- [x] I confirm that I am using English to submit this report ([Language Policy](https://github.com/infiniflow/ragflow/issues/5910)).
- [x] Non-english title submitions will be closed directly ( 非英文标题的提交将会被直接关闭 ) ([Language Policy](https://github.com/infiniflow/ragflow/issues/5910)).
- [x] Please do not modify this template :) and fill in all the required fields.

### RAGFlow workspace code commit ID

9799f33549ec45415a14689882e90faa481b9d04

### RAGFlow image version

9799f335(v0.25.6)

### Other environment information

```Markdown
- Hardware parameters: N/A (reproducible via unit test / API call)
- OS type: Linux
- Others: Affects any deployment using `async_ask()` — e.g. `POST /api/v1/searchbots/ask` and search completion SSE endpoints
```

### Actual behavior

`async_ask()` in `api/db/services/dialog_service.py` calls `KnowledgebaseService.get_by_ids(kb_ids)` and immediately accesses `kbs[0].tenant_id` without checking whether any knowledge bases were returned.

When `kb_ids` is empty, or every ID is deleted/unknown, `get_by_ids()` returns an empty list and the function raises:

`IndexError: list index out of range`

This is caught by the route's generic exception handler and returned to the client as HTTP 500 with a message like `**ERROR**: list index out of range` in the SSE stream.

### Expected behavior

When no knowledge bases are found, `async_ask()` should return a graceful error in the SSE stream instead of crashing.

`gen_mindmap()` in the same module already handles this correctly:

```python
kbs = KnowledgebaseService.get_by_ids(kb_ids)
if not kbs:
    return {"error": "No KB selected"}

### Steps to reproduce

```Markdown
1. Deploy or run RAGFlow from commit `9799f335` (v0.25.6) or current `main`.

2. Obtain a valid API token for the embedded searchbot endpoints.

3. Send a POST request to `/api/v1/searchbots/ask` with a valid `question` and an invalid or deleted `kb_ids` value.

   Example request body:

   {"question": "What is RAGFlow?", "kb_ids": ["nonexistent-kb-id"]}

4. Observe the response.

   Actual result: HTTP 500, and the SSE stream contains an error like `**ERROR**: list index out of range`.

   Expected result: a graceful SSE error such as `**ERROR**: No KB selected`, without an unhandled exception.

---

Alternative code-level reproduction:

1. Open `api/db/services/dialog_service.py` and locate `async_ask()`.

2. After this line:

   kbs = KnowledgebaseService.get_by_ids(kb_ids)

   the code accesses `kbs[0].tenant_id` without checking whether `kbs` is empty.

3. Call `async_ask()` with empty `kb_ids`:

   async_ask(question="What is RAGFlow?", kb_ids=[], tenant_id="tenant-1")

4. Result: `IndexError: list index out of range`.
```

### Additional information

Affected code path:

File: api/db/services/dialog_service.py
Function: async_ask()

After this line:
  kbs = KnowledgebaseService.get_by_ids(kb_ids)

The code immediately uses:
  kbs[0].tenant_id

If get_by_ids() returns an empty list (empty kb_ids, deleted KBs, or invalid IDs), this raises:
  IndexError: list index out of range

---

Affected endpoints:

- POST /api/v1/searchbots/ask (api/apps/restful_apis/bot_api.py)
- Search completion SSE streams (api/apps/restful_apis/search_api.py)

Callers validate that kb_ids is present in the request, but not that the IDs still exist in the database.

---

Existing correct pattern in the same file:

gen_mindmap() already guards against this:

  kbs = KnowledgebaseService.get_by_ids(kb_ids)
  if not kbs:
      return {"error": "No KB selected"}

async_ask() does not have an equivalent guard.

---

Suggested fix:

  if not kbs:
      yield {"answer": "**ERROR**: No KB selected", "reference": {}, "final": True}
      return

---

Error seen by client (via route exception handler):

  {"code": 500, "message": "list index out of range", "data": {"answer": "**ERROR**: list index out of range", "reference": []}}

I have a patch and regression tests ready on branch fix/async-ask-empty-kb-guard and can open a PR after this issue is filed.

### PR DESCRIPTION
**Pull request by bohdansolovie**
Fixes #15529 .

### Problem

`async_ask()` accessed `kbs[0]` without verifying that `KnowledgebaseService.get_by_ids()` returned any knowledge bases. Empty or stale `kb_ids` raised `IndexError`, which surfaced as HTTP 500 on search/bot SSE endpoints.

### Fix

- Add an early guard when `kbs` is empty, yielding a final SSE error event (consistent with `gen_mindmap()` in the same module).
- Add regression tests for empty `kb_ids` and deleted/invalid KB IDs.

### Test plan

- [ ] `pytest test/unit_test/api/db/services/test_dialog_service_final_answer.py -k "async_ask_empty or async_ask_stale"`
- [ ] Manual: `POST /api/v1/searchbots/ask` with invalid `kb_ids` returns SSE error, not HTTP 500

### CODE REVIEW

**Comment by bohdansolovie** [comment]
@yingfeng
Could you please review?

**Comment by bohdansolovie** [comment]
@wangq8 , @Haruko386 , @JinHai-CN @yingfeng @xugangqiang
Please review this PR

File: api/db/services/dialog_service.py [hunk 1]
```diff
@@ -1533,6 +1533,10 @@ async def async_ask(question, kb_ids, tenant_id, chat_llm_name=None, search_conf
     include_reference_metadata, metadata_fields = _resolve_reference_metadata(search_config)
 
     kbs = KnowledgebaseService.get_by_ids(kb_ids)
+    if not kbs:
+        yield {"answer": "**ERROR**: No KB selected", "reference": {}, "final": True}
```
Comment (xugangqiang) [inline api/db/services/dialog_service.py]: No KB selected ? Or the KB selected is not valid?

Comment (bohdansolovie) [reply api/db/services/dialog_service.py]: I have solved this issue.
@xugangqiang Please check again.

File: api/db/services/dialog_service.py [hunk 2]
```diff
@@ -1533,6 +1533,14 @@ async def async_ask(question, kb_ids, tenant_id, chat_llm_name=None, search_conf
...
 
     kbs = KnowledgebaseService.get_by_ids(kb_ids)
+    if not kbs:
+        if not kb_ids:
+            error = "**ERROR**: No KB selected"
+        else:
+            error = "**ERROR**: The selected KB is not valid"
+        yield {"answer": error, "reference": {}, "final": True}
+        return
+
```
Comment (xugangqiang) [inline api/db/services/dialog_service.py:1610]: Is it possible to test this on UI or curl?  We'd like to see what will happen with this fix.
